### PR TITLE
[KOGITO-9982] Add Persistence field in SonataFlow CRD

### PR DESCRIPTION
**Pull request by jordigilh**
* Adds the Persistence field in the SonataFlow CRD
* Adds the implementation to handle PostgreSQL persistence type 
* If `Persistence` field is empty or nil, no persistence is defined.

@tchughesiv @wmedvede @ricardozanini PTAL.

### CODE REVIEW

File: controllers/profiles/common/persistence.go
Comment (ricardozanini) [inline controllers/profiles/common/persistence.go:43]: Can you move these hardcoded strings and integers to a private const in this file?

File: api/v1alpha08/sonataflow_types.go
Comment (masayag) [inline api/v1alpha08/sonataflow_types.go:662]: @ricardozanini shouldn't the decision about persistency of the runtime be determined by the Platform?

Comment (jordigilh) [reply api/v1alpha08/sonataflow_types.go:662]: The idea is that this field can be used to override the platform's default persistence, or define one in case it does not exist. The platform's persistence field does not exist yet, it will be implemented in [KOGITO-9980](https://issues.redhat.com/browse/KOGITO-9980)

Comment (masayag) [reply api/v1alpha08/sonataflow_types.go:662]: Ok, that makes sense then.

**Review by jakubschwan — COMMENTED** [review]
Changes looks good to me. Please also check and resolve failing PR checks.

**Comment by domhanak** [comment]
@jordigilh another rebase should solve the test issues. Other than that it looks good.

**Comment by domhanak** [comment]
Looks like the added test case is not working here on CI:
Needs to be stable to merge,  I am trying to reproduce the test failure locally.
```
Validate the persistence  when deploying a SonataFlow CR with PostgreSQL persistence [It] defined in the workflow from an existing kubernetes service as a reference
/home/runner/work/incubator-kie-kogito-serverless-operator/incubator-kie-kogito-serverless-operator/test/e2e/persistence_test.go:123

  [FAILED] Timed out after 600.001s.
  Expected success, but got an error:
      <*errors.errorString | 0xc00007a2a0>: 
      kubectl wait pod -n test-546 -l sonataflow.org/workflow-app=callbackstatetimeouts --for condition=Ready --timeout=5s failed with error: (exit status 1) error: timed out waiting for the condition on pods/callbackstatetimeouts-7bf57f9968-5nfsl
      
      {
          s: "kubectl wait pod -n test-546 -l sonataflow.org/workflow-app=callbackstatetimeouts --for condition=Ready --timeout=5s failed with error: (exit status 1) error: timed out waiting for the condition on pods/callbackstatetimeouts-7bf57f9968-5nfsl\n",
      }
```

**Comment by jordigilh** [comment]
> Looks like the added test case is not working here on CI: Needs to be stable to merge, I am trying to reproduce the test failure locally.
> 
> ```
> Validate the persistence  when deploying a SonataFlow CR with PostgreSQL persistence [It] defined in the workflow from an existing kubernetes service as a reference
> /home/runner/work/incubator-kie-kogito-serverless-operator/incubator-kie-kogito-serverless-operator/test/e2e/persistence_test.go:123
> 
>   [FAILED] Timed out after 600.001s.
>   Expected success, but got an error:
>       <*errors.errorString | 0xc00007a2a0>: 
>       kubectl wait pod -n test-546 -l sonataflow.org/workflow-app=callbackstatetimeouts --for condition=Ready --timeout=5s failed with error: (exit status 1) error: timed out waiting for the condition on pods/callbackstatetimeouts-7bf57f9968-5nfsl
>       
>       {
>           s: "kubectl wait pod -n test-546 -l sonataflow.org/workflow-app=callbackstatetimeouts --for condition=Ready --timeout=5s failed with error: (exit status 1) error: timed out waiting for the condition on pods/callbackstatetimeouts-7bf57f9968-5nfsl\n",
>       }
> ```

Fixed 😄

**Comment by jordigilh** [comment]
@ricardozanini ready to be merged. Please note that I had to remove the `DEBUG=true` env value in the job because it caused the `kubectl` commands to dump extra log entries and that caused errors when parsing the output as it did not match the expected output.

**Comment by ricardozanini** [comment]
@jordigilh many thanks! we Will wait for QE verification first with @domhanak

**Review by domhanak — APPROVED** [review]

**Comment by ricardozanini** [comment]
> Please note that I had to remove the `DEBUG=true` env value in the job because it caused the `kubectl` commands to dump extra log entries and that caused errors when parsing the output as it did not match the expected output.

Thanks! I've removed this env var too in an opened PR! 😄

File: controllers/platform/services/services.go
Comment (ricardozanini) [inline controllers/platform/services/services.go:139]: Can you move these to private consts?

File: controllers/profiles/common/object_creators_test.go
Comment (ricardozanini) [inline controllers/profiles/common/object_creators_test.go:266]: nit; are you using this var outside the `TestMergePodSpec_OverrideContainers_WithPostgreSQL_and_ServiceRef` test scope?

File: controllers/profiles/common/persistence/postgreSQL.go
Comment (ricardozanini) [inline controllers/profiles/common/persistence/postgreSQL.go:87]: I'd rather keep all these hardcoded strings in private const.

File: test/e2e/persistence_test.go
Comment (ricardozanini) [inline test/e2e/persistence_test.go:38]: Nice one!

File: test/e2e/workflow_test.go
Comment (ricardozanini) [inline test/e2e/workflow_test.go:46]: 👍

File: test/testdata/persistence/workflow/by_service/02-sonataflow_platform.yaml
Comment (ricardozanini) [inline test/testdata/persistence/workflow/by_service/02-sonataflow_platform.yaml:24]: @wmedvede I think it's worth adding those to the images, so we keep them in the cache to facilitate persistence use cases. wdyt?

File: test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml
Comment (ricardozanini) [inline test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml:43]: Just as a quick note; we have a JIRA to implement this in case the workflow requires persistence: https://issues.redhat.com/browse/KOGITO-9840

Can you please add this link with `TODO:` in this file? 🙏

File: test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml
Comment (ricardozanini) [inline test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml:41]: I guess we can use https://catalog.redhat.com/software/containers/ubi9/ubi-micro/615bdf943f6014fa45ae1b58 instead. IIRC `coreutils` includes cat.

It's 7 vs 36 MB. Quite a difference in storage.

**Review by ricardozanini — COMMENTED** [review]
I just left some minor comments if you don't mind. Otherwise looks good! Many thanks!

**Comment by ricardozanini** [comment]
@tchughesiv can you please review it too?

**Review by tchughesiv — APPROVED** [review]
lgtm

File: controllers/profiles/common/object_creators_test.go
Comment (jordigilh) [reply controllers/profiles/common/object_creators_test.go:266]: No. Why?

File: controllers/profiles/common/persistence/postgreSQL.go
Comment (jordigilh) [reply controllers/profiles/common/persistence/postgreSQL.go:87]: Moved to private constants

File: controllers/profiles/common/object_creators_test.go
Comment (ricardozanini) [reply controllers/profiles/common/object_creators_test.go:266]: I'd keep it within the test scope then.

File: test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml
Comment (jordigilh) [reply test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml:41]: Changed to use `ubi-micro:latest` in this manifest

File: test/testdata/persistence/workflow/by_service/02-sonataflow_platform.yaml
Comment (wmedvede) [reply test/testdata/persistence/workflow/by_service/02-sonataflow_platform.yaml:24]: I think we need to assess here, because the persistence is a sort of all or nothing. I mean, if we add the org.kie.kogito:kogito-addons-quarkus-persistence-jdbc:999-SNAPSHOT to the swf-builder, then, every w the workflow must use persistence, otherwise we'll have strart-up failures.

O think we have the following scenarios:

1) The images is produced externally:
The workflow must come already built with the proper extensions incorporated in the produced image, and users might decide in a per workflow basis to include the persistence related addons. Finally the operator just relies on the workflow CRD con connect with the DB.

2) The Image is produced by the operator managed Kaniko or OpenShift based builds, here

But we are dealing with the persistence as an optional thing, so, it could happen the a workflow CRD with no configured persistence might come.

File: test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml
Comment (wmedvede) [reply test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml:44]: General comment here guys  @jordigilh @ricardozanini that doesn't block this PR.

I think we need to assess here, because the persistence is a sort of all or nothing.

I mean, if we add the **org.kie.kogito:kogito-addons-quarkus-persistence-jdbc:999-SNAPSHOT** to the **kogito-swf-builder**, then, every workflow image we produce must use persistence, otherwise we'll have runtime start-up failures.
Additionally, we have the following scenarios:

**1) The image is produced externally:**
The workflow must come already built with the proper extensions incorporated in the produced image, and users might decide in a per-workflow basis to include the persistence related addons.
The operator just relies on the workflow CRD con connect with the DB, and just produce the ENV vars as Jordi is doing. 
If the image is not ok, it's not the operator problem. Which make sense in my opinion.

(right now, AFAIK Parodos team is using the kogito-swf-builder series to produce their images, and I think they might eventually produce the images by adding the persistence related extensions in a per-workflow basis,
or always add them if they always want persistence)

**2) The Image is produced by the operator, and thus, managed by Kaniko or OpenShift based builds, etc.**
In these cases, we could eventually manage to add the persistence extensions somehow dynamically, since in the end, for example with the Kaniko build we are producing a sonataflow-my-workflow-builder CM and here we can maybe manage to add these extensions.

But, since Parodos is working with 1), I think they should add those extensions in their images build system.

**Review by wmedvede — APPROVED** [review]
LGTM.

I have added a general comment regarding the images generation that I think make sense, take a look please

**Comment by wmedvede** [comment]
Hi @jordigilh , it looks like we have some merge conflicts now, would you mind take a look please.

File: test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml
Comment (ricardozanini) [reply test/testdata/persistence/workflow/by_service/03-sonataflow_callbackstatetimeouts.sw.yaml:44]: [we are going to add the concept of traits](https://issues.redhat.com/browse/KOGITO-8255), that will handle use cases like this. So even if produced externally, the operator will know if the workflow depends on persistence or not and act accordingly.

**Comment by jordigilh** [comment]
@ricardozanini PR is green. Can we merge it?

**Comment by ricardozanini** [comment]
Let's wait for @wmedvede approval. I'll also take a look on it tomorrow.

**Comment by ricardozanini** [comment]
@jordigilh can you please open a task to update the docs too? Thank you!

**Pull request merged**
